framework/src/main/java/org/checkerframework/framework/source/CompositeChecker.java

```diff
@@ -0,0 +1,684 @@
+package org.checkerframework.framework.source;
+
+import com.google.common.collect.ImmutableSet;
+import com.sun.source.tree.CompilationUnitTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.util.TreePath;
+import com.sun.tools.javac.processing.JavacProcessingEnvironment;
+import com.sun.tools.javac.util.Context;
+import com.sun.tools.javac.util.Log;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeSet;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.TypeElement;
+import javax.tools.Diagnostic;
+import org.checkerframework.checker.interning.qual.FindDistinct;
+import org.checkerframework.checker.interning.qual.InternedDistinct;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.checker.signature.qual.ClassGetName;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.common.reflection.MethodValChecker;
+import org.checkerframework.framework.util.TreePathCacher;
+import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.InternalUtils;
+import org.plumelib.util.ArrayMap;
+import org.plumelib.util.CollectionsPlume;
+
+/**
+ * An abstract {@link SourceChecker} that provides support for subcheckers and manages their
+ * messages. It doesn't require an extending checker to provide a qualifier hierarchy, type factory
+ * or visitor. Hence, it can be used similar to an {@link AggregateChecker}. However, the
+ * subcheckers of a CompositeChecker can cooperate and communicate, which is not possible using an
```

Comment:
The immediate subcheckers of an Aggregate Checker can't communicate, but if a subchecker has subcheckers then those checkers can communicate.    (We test this in `org.checkerframework.checker.testchecker.NestedAggregateChecker`.) So the Resource Leak Checker in this pull request should be able to be an Aggregate Checker.  

However, when I tried this, `-AwarnUnneededSuppressions` did not work correctly.  I think this is a bug in the Aggregate Checker.  The Aggregate Checker has been hard to maintain and with this change doesn't have a clear use case.  I would propose deprecating the Aggregate Checker and using the Composite Checker instead. (Actually, I think we could just move all the code in the Composite Checker into Source Checker.  The only thing in Composite Checker would be the implementation of  `createSourceVisitor()`.

Reply:
If we move the code from the composite checker into the source checker, do you want to maintain that the subcheckers are BaseType checkers, the way it is now?

Do you want to make these changes yourself within this PR?

Reply:
> If we move the code from the composite checker into the source checker, do you want to maintain that the subcheckers are BaseType checkers, the way it is now?

No.

> Do you want to make these changes yourself within this PR?

I'll do it in a separate PR.